starlite/contrib/sqlalchemy/repository.py

```diff
@@ -0,0 +1,540 @@
+
+__all__ = (
+    "SQLAlchemyRepository",
+    "ModelT",
+)
+
+T = TypeVar("T")
+ModelT = TypeVar("ModelT", bound="base.Base | base.AuditBase")
+SQLARepoT = TypeVar("SQLARepoT", bound="SQLAlchemyRepository")
+SelectT = TypeVar("SelectT", bound="Select[Any]")
+RowT = TypeVar("RowT", bound=Tuple[Any, ...])
+
+
+@contextmanager
+def wrap_sqlalchemy_exception() -> Any:
+    """Do something within context to raise a `RepositoryError` chained
+    from an original `SQLAlchemyError`.
+
+        >>> try:
+        ...     with wrap_sqlalchemy_exception():
+        ...         raise SQLAlchemyError("Original Exception")
+        ... except RepositoryError as exc:
+        ...     print(f"caught repository exception from {type(exc.__context__)}")
+        ...
+        caught repository exception from <class 'sqlalchemy.exc.SQLAlchemyError'>
+    """
+    try:
+        yield
+    except IntegrityError as exc:
+        raise ConflictError from exc
+    except SQLAlchemyError as exc:
+        raise RepositoryError(f"An exception occurred: {exc}") from exc
+
+
+class SQLAlchemyRepository(AbstractRepository[ModelT], Generic[ModelT]):
+    """SQLAlchemy based implementation of the repository interface."""
+
+    def __init__(
+        self, *, session: AsyncSession, base_select: Select[tuple[ModelT]] | None = None, **kwargs: Any
+    ) -> None:
+        """Repository pattern for SQLAlchemy models.
+
+        Args:
+            session: Session managing the unit-of-work for the operation.
+            base_select: To facilitate customization of the underlying select query.
+            **kwargs: Additional arguments.
+
+        """
+        super().__init__(**kwargs)
+        self.session = session
+        self.statement = base_select or select(self.model_type)
+
+    async def add(self, data: ModelT) -> ModelT:
+        """Add `data` to the collection.
+
+        Args:
+            data: Instance to be added to the collection.
+
+        Returns:
+            The added instance.
+        """
+        with wrap_sqlalchemy_exception():
+            instance = await self._attach_to_session(data)
+            await self.session.flush()
+            await self.session.refresh(instance)
+            self.session.expunge(instance)
+            return instance
+
+    async def add_many(self, data: list[ModelT]) -> list[ModelT]:
+        """Add Many `data` to the collection.
+
+        Args:
+            data: list of Instances to be added to the collection.
+
+        This function has an optimized bulk insert based on the configured SQL dialect:
+        - For backends supporting `RETURNING` with `executemany`, a single bulk insert with returning clause is executed.
+        - For other backends, it does a bulk insert and then selects the inserted records
+
+        Returns:
+            The added instances.
+        """
+        data_to_insert: list[dict[str, Any]] = [v.to_dict() if isinstance(v, self.model_type) else v for v in data]  # type: ignore
```

Comment:
can we not just pass the list of declarative models straight through to the bulk insert? I.e., do we need this dict conversion?

Also, why the `isinstance` check? The method signature says that the data is a list of `ModelT`, so anything else would be a type error, right?

Also, we can probably relax that data type back to `Collection` or `Iterable` or smth.

Reply:
This is the way it's done in the SQLAlchemy docs: https://docs.sqlalchemy.org/en/20/orm/queryguide/dml.html#orm-bulk-insert-statements. 

This though raises the question of the signature of this function. 

To me it would make most sense if either this function operates on raw data and not instances when using bulk inserts like this, or or it operates on instances and just uses the session methods for this (e.g. `add_all`). 

The current way mixes both, which is a bit confusing. If those are newly created instances, they'd also not belong to a session, so I'm not sure if `expunge` is even necessary? 

It also results in:

- Construction of transient instances
- Conversion of those instances into dicts
- Creation of new, non-transient but otherwise identical, instances from `session.scalars`  

which seems like an unnecessary round trip.

Reply:
> This is the way it's done in the SQLAlchemy docs

Indeed, and having just tested, it definitely doesn't work with instances.

> To me it would make most sense if either this function operates on raw data and not instances when using bulk inserts like this, or or it operates on instances and just uses the session methods for this (e.g. add_all).

The general repository pattern is to deal in ORM objects, and it would be nice to remain consistent with that. However, parsing out of bytes, into some ser/de model, into an orm model, to finally convert to dicts, and back again into orm models is indeed a very round-about means to the end.

If I were very concerned with efficiency (i.e., collection size * number of attributes) I'd probably have a typed dict representation of my data, and go straight from msgspec to that, and that case, I would certainly appreciate the repository having methods to support.

Maybe we should mirror the sqlalchemy api for this stuff then - a bulk method for mappings, and a separate one for orm objects. It looks as though the `add_all` method should still participate in the benefits of the improved 'returning' support in 2.0: https://docs.sqlalchemy.org/en/20/changelog/whatsnew_20.html#benchmarks. Our recommendation could be that if performance isn't a major factor, and collection sizes aren't very big, then use the ORM one, and where performance is critical, we direct users to the mappings method, and then suggest not parsing into ORM objects in the first place.

Reply:
> Maybe we should mirror the sqlalchemy api for this stuff then - a bulk method for mappings, and a separate one for orm objects.

> Our recommendation could be that if performance isn't a major factor, and collection sizes aren't very big, then use the ORM one, and where performance is critical, we direct users to the mappings method, and then suggest not parsing into ORM objects in the first place.

That sounds like the proper solution to this. Especially since `session.add_all` now uses `RETURNING` internally, performance should be still be quite good.

Reply:
I'd also argue that if there is a concern for performance, `sqlite` likely won't be used and the `insert` `returning` will be executed.

Reply:
I also would need to verify that `add_all` doesn't increase the overhead too much.  I have used the bulk insert method above to insert thousands of objects at a time, and, in the past, I had issues with pretty much every other sqlalchemy method.  (This was obviously be fore the 2.0 change)